extensions/ql-vscode/src/query-history.ts

```diff
@@ -173,33 +173,47 @@ export class HistoryTreeDataProvider extends DisposableObject implements TreeDat
 
     // Populate the icon and the context value. We use the context value to
     // control which commands are visible in the context menu.
-    let hasResults;
+    treeItem.iconPath = this.getIconPath(element);
+    treeItem.contextValue = await this.getContextValue(element);
+
+    return treeItem;
+  }
+
+  private getIconPath(element: QueryHistoryInfo): ThemeIcon | string {
     switch (element.status) {
       case QueryStatus.InProgress:
-        treeItem.iconPath = new ThemeIcon('sync~spin');
-        treeItem.contextValue = element.t === 'local' ? 'inProgressResultsItem' : 'inProgressRemoteResultsItem';
-        break;
+        return new ThemeIcon('sync~spin');
       case QueryStatus.Completed:
         if (element.t === 'local') {
-          hasResults = await element.completedQuery?.query.hasInterpretedResults();
-          treeItem.iconPath = this.localSuccessIconPath;
-          treeItem.contextValue = hasResults
+          return this.localSuccessIconPath;
+        } else {
+          return this.remoteSuccessIconPath;
+        }
+      case QueryStatus.Failed:
+        return this.failedIconPath;
+      default:
+        assertNever(element.status);
+    }
+  }
+
+  private async getContextValue(element: QueryHistoryInfo): Promise<string> {
```

Comment:
Can we extract this method out of the class so it can be tested separately? It doesn't seem like it uses `this` anywhere.

Reply:
You're right it's only `getIconPath` that relies on `this`, so we could pull out just `getContextValue`. I suppose doing it for one method is better than for zero methods.

Reply:
Ok, I've tried and I'm struggling quite a lot. The problem is that `getContextValue` depends on `element.completedQuery?.query.hasInterpretedResults()` and this calls into `QueryEvaluationInfo` and generally hits two problems:

The real implementation of the method relies on disk state, and the query history info tests aren't really set up for that. We could work around that passing in a helper method to `getContextValue`. It's a bit ugly but it would work. However there are deeper problems that than.

We don't do a good job of creating the right types in our tests. In this case `createMockLocalQueryInfo` doesn't actually create a `LocalQueryInfo`. It casts a random object to that type, but in reality `LocalQueryInfo` is a class so you can't just do this. This means when we call `element.status` it returns undefined because although `element.t === 'local'` it is not an instance of the `LocalQueryInfo` type, so the `status` method is not defined. I had a look at whether I could instantiate the class easily but it looks quite complicated. We do do some of this in `query-history.test.ts` but at this point things are getting very complicated and wide-ranging.

I'd like to merge this change as it is and then look at improving things separately.

Reply:
Fwiw, I'm going to introduce an improved version of the factory which does create a proper object. I need it in order to make sorting tests work. So you might be able to do this after that's merged.

Reply:
This has now been merged: https://github.com/github/vscode-codeql/pull/1710